Improvements to complex vouchers

jniles opened the pull request:
This commit refactors the Advanced Journal Vouchers to meet all the requirements stated in #2196.

### Why is this important?
As a reminder of the motivation behind #2196 which may not have been clearly articulated (or too long ago now to remember), the reason a refactor of this code was important was two-fold:
  1. We hoped to use the same or similar components to what was used in the transaction edit modal for journal vouchers to both reuse code between the two, and also reduce the differences in what users had to learn to make both work.  I partially achieved code reuse in this rewrite, but ended up removing redundancies more than reusing code.
  2. We also observed that as more transactions were added to the database, there was no reasonable way of creating a quick interface for linking entities or references to vouchers.  The old way required loading _every single transaction_ for cash, vouchers, and invoices, then searching for the specific one by reference, then clicking on it.  This was slow (at risk of crashing small servers in practice) and rarely done by both us the developers and users.  We never recommended it after we implemented it.

These reasons above were the background of #2196.

### What does this PR do?
In addition to satisfying the objectives laid out in the issue, this PR cleans up a lot of cruft in the complex vouchers form service.  Notably, the modals to find references and entities have been removed, and the voucher tools updated to deal solely with uuids instead of having to create debtor/creditor references.  This makes the complex vouchers more performant, easier to use, and code clearer.

Here is what it looks like:

<img width="1394" alt="image" src="https://github.com/IMA-WorldHealth/bhima/assets/896472/ccc77398-e879-49aa-aa94-9d2fe27ece14">

### What is next?
This is a pretty significant change and may need some retraining of users at installation sites.  I'm posting this as a draft to get comments from the team before turning it into a full PR.  It's been 4 years in the making 😆 !

I still believe now as I did 4 years ago that this is the best way to make Complex Vouchers more usable.  The PR also perfectly sets up the codebase to implement https://github.com/IMA-WorldHealth/bhima/issues/2101 in a future PR without too much issue. Combined, these will make Complex Vouchers more useful for modeling complicated transactions.

It also paves the way for https://github.com/IMA-WorldHealth/bhima/issues/5936 to be implemented as well.

Closes #2196.

jniles committed:
feat: add text lookups for entity/records

This commit adds helper functions to the shared finance file to lookup
entity and record uuids based on the human readable text identifiers.
These are useful when editing vouchers or transactions in which the user
specifies the text, but the underlying link must be made as a UUID.

jniles committed:
feat(vouchers): add lookup of hrEntity to API

This commit adds the ability to create vouchers with entity_uuids by
specifying an hrEntity that is automatically linked via helper
functions.  Currently, if an hrEntity does not match, the API simply
ignores it.  In the long run, we might want to block the user if we
cannot find a match for the entity specified - this would require
flipping to a series of db.one()s or comparing unique array lengths.
db.one() would be less efficient (more DB calls), but arguably simpler
to read and maintain.

jniles committed:
feat(vouchers): lookup document_uuid by hrRecord

In the same way as the entity can be looked up by specifying the
hrEntity key, the document linked can also be looked up by specifying an
hrRecord key.  This document is matched against the document_map table
and the uuid is extracted from there.

jniles committed:
fix: join on debtor/creditor for entity lookup

Instead of implementing #3514 which would require massive migration
scripts, we simply JOIN the debtor/creditor on the entity_map.  It less
performant, but doesn't require a breaking change.

jniles committed:
fix(vouchers): translate negative value error

Fixes a typo that caused the negative value error not to display.

Closes #3507.

jniles committed:
chore(i18n): improve form translation

jniles committed:
feat: add bhEntityTypeahead component

This commit adds the bhEntityTypeahead component for looking up entities
based on their reference in real time.  It is used on the complex
voucher page to provide real-time filtering of the entity name.

jniles committed:
feat: add bhRecordTypeahead component

The bhRecordTypeahead component is similar to the bhEntityTypeahead
component by providing real-time access to the records in the database.
The current syntax/performance could be improved later by having
separate JOIN queries of document_map onto each individual record
tables.

jniles committed:
chore(vouchers): transition to typeaheads in grid

This commit begins transitioning the complex vouchers towards using a
typehead in the ui-grid.  The previous version using links was untenable
for environments with thousands or hundreds of thousands of records.
The typeahead model allows users to fluidly select records by simply
typing.

jniles committed:
feat(vouchers): make record typeahead drop left

This commit makes the bhRecord typeahead drop left in all cases.  This
solves the constraint of space on the complex voucher grid.

jniles committed:
fix(vouchers): remove the recover cache button

This commit removes the recover cache button as requested in #2196.

Close #2196.

jniles committed:
fix(vouchers): tools populate entity/records

Vouchers tools can now set the entity and records for each of the items
in the voucher grid.

jniles committed:
feat(bhEntityTypeahead): cancel pending requests

This commit adds the ability to cancel pending requests if new ones are
introduced.  It should produce some performance improvements.

jniles committed:
feat(bhRecordTypeahead): add validation classes

Adds validation to the inputs in the typeaheads to make sure they are
accurately reflective of their validation state(s).

jniles committed:
fix(vouchers): ensure that entity is picked up

This commit ensures that the entity is formatted correctly in both the
SQL read request and the voucher create() call from the client.

jniles committed:
fix: remove console.log()

jniles committed:
fix: remove double 'dm' references

Fixes an SQL bug in the merging of the PR.

jniles committed:
fix(voucher): voucher-tools do not null entity field

This commit ensures that voucher tools to do null the entity field,
updating the patient support tool to support the latest standard.

jniles committed:
chore: rm findReferenceModel, findEntityModal

This commit removes the findReferenceModal and findEntityModals since
they are no longer used.  Also cleans up all associated code and
unnecessary formatting.

Closes #2196.

mbayopanda commented on client/src/js/components/bhEntityTypeahead.js:
```diff
@@ -0,0 +1,58 @@
+  };
+
+  // cancels all pending requests
+  function cancelInProgressRequests() {
```
👍🏽

mbayopanda commented on client/src/modules/templates/bhAccountTypeaheadInline.html:
```diff
@@ -2,7 +2,7 @@
     class="form-group"
     style="margin-bottom:0;"
-    ng-class="{ 'has-error' : AccountForm.account.$error && AccountForm.$$parentForm.$submitted }">
+    ng-class="{ 'has-error' : AccountForm.account.$error && AccountForm.$submitted }">
```
👍🏽

mbayopanda commented on server/controllers/finance/vouchers.js:
```diff
@@ -358,16 +389,14 @@ function createVoucher(voucherDetails, userId, projectId) {
 
   // Only for Employee Salary Paiement
-  if (voucherType === 7) {
+  if (voucherTypeId === SALARY_PAYMENT_VOUCHER_TYPE_ID) {
```
👍🏽

mbayopanda reviewed (approved):
This is an improvement in the UX and in the code.

The UI and the feature are working as expected and fine, and also seeing the entity and reference in the receipt is another point of stability of the feature.

There are no comments on the code, it's excellent.

You have just to update the `test/data.sql` to add the fiscal year 2024

jmcameron reviewed (changes requested):
I have not read the code in detail yet.
Basic tests pass.  But the end-to-end tests related to vouchers fail (more on this below).
When creating a simple voucher, the dropdown for the "Voucher Type" does not seem to work.

To run just the voucher end-to-end tests, temporarily add the following line to package.json:
`"test:e2e-test": "cross-env SUITE=\"end-to-end\" TESTS=\"test/end-to-end/vouchers/\" TEST_NUM=8 ./sh/test-e2e.sh",`
and run the test with:  `npm run test:e2e-test`

A few notes on the end-to-end voucher tests:
- in `test/end-to-end/couchers/complex.spec.js`, 
   - the const `voucher` needs to be updated with suitable values for Debtor/creditor and Reference
- in `test/end-to-end/vouchers/voucherRowClass.js`
   - reference()` and `entity()` need to be updated since they no longer use popups.  I was not clear on how this was supposed to work.
- The simple voucher test may need updating based on how the dropdown issue is resolved.

jniles committed:
test(voucher): unbreak voucher tests

Updates the test harness for the new voucher creation syntax.

jmcameron reviewed (changes requested):
There still seem to be problems on my local server.
- In Simple Voucher form, the "Transaction Type" drop-down does not work (for me).  It is a drop-down, correct?  not a type-ahead?  BTW, this works fine in master.
- The simple voucher end-to-end tests fail:
   - `test/end-to-end/vouchers/simple.spec.js:36:3 › Simple Vouchers › can create a simple voucher`
   - You can try this using the method suggested in https://github.com/IMA-WorldHealth/bhima/pull/7386#pullrequestreview-1826701649.  Or run `test:e2e-8`.

jniles committed:
fix(vouchers): unbreak simple vouchers

Re-fixes simple vouchers.

jmcameron commented on test/end-to-end/shared/TestUtils.js:
```diff
@@ -370,7 +370,7 @@ module.exports = {
 
     // select the item of the dropdown menu matching the label
-    return anchor.locator(`.dropdown-menu > [role="option"] >> text="${label}"`).click();
+    await anchor.locator(`.dropdown-menu > [role="option"] >> text="${label}"`).click();
```
This  function is advertised as returning a promise.  It used to do that with the final 'return...'.   Now it does not do that.  Was there a reason for this change?

jniles replied:
Nope, fixed.

jmcameron commented on server/config/routes.js:
```diff
@@ -1064,6 +1066,11 @@ exports.configure = function configure(app) {
   app.get('/lots/generate_barcodes/:number', lots.generateBarcodes);
 
+  app.get('/finance/entities', financeShared.lookupFinancialEntity);
+  app.get('/finance/entities/:uuid', financeShared.lookupFinancialEntityByUuid);
```
Are the tests for the /finance/entity and /finance/record API endpoints?

jniles replied:
Just added some!  This was a good catch too, because I found an unexpected behavior around LIMITs.

jmcameron reviewed (changes requested):
Simple vouchers work now.
However, I am still confused about how the 'entity' and 'record' typeahead fields are supposed to work in complex vouchers.   When I try to into them, nothing happens.  Is that the correct behavior (with bhima_test) because no entity/records are available, or am I missing something?  Operator error?

The code looks good.   I made a few comments on things I noticed.

I noticed the a very minor 'nit':  In the voucher registry, the column header for the action menu should not appear in the Column selector.  This is due to setting the `displayName` for the action column to '...' instead of '' (see below).   If you don't mind, please do a global search and replace to change `displayName : '...'` to 'displayName : ''`.  I noticed 5 occurences for this and they are all for action items.  Note that these are not a big deal in the main grid displays, but using '...' means that they get added to the Column selector, which does not make a lot of sense since you always want them, and a checkbox with ... beside it is not very informative!
![image](https://github.com/IMA-WorldHealth/bhima/assets/62145/fb99fc09-ab91-499f-8f89-f6f8e8e1c830)

jniles committed:
test(finance): integration for entities/records

Adds some integration tests for the routes:
 1. finance/entities
 2. finance/records

to make sure they conform to their API even if modified.

jmcameron reviewed (approved):
It all seems to work now.  I suggest merging with master and rebasing.  Also, if you could fix the minor "nit" I mentioned at the end of [this previous comment](https://github.com/IMA-WorldHealth/bhima/pull/7386#pullrequestreview-1848784760), please do that, or add a new issue so we can fix it later.

Regarding the `entity` and `reference` fields in the complex voucher form:  I found it usable, once I realized that that you have to type in 3 characters to get it to show choices.  But it took some fumbling around to guess what the possible prefixes are.   I'm not sure how to improve that since you have to know what code prefixes are possible for each field before you start typing.  However, since the developer/database would usually(?) know what the possible prefix code options are, perhaps when you click in the input field you get a modal dialog with the possible options.  Click on one and a drop-down appears with the available items. Click on the one you want and it fills out the original input form.   Perhaps an issue for later?   Also, do we have any hints elsewhere to indicate when a field requires 3 characters to be typed in before anything happens that could be applied here?   Or do we assume that a person with training and experience would know what to do?
